[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69854
* __->__ #69853
* #69852

We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

swolchok committed:
[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)

We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

suo commented on torch/csrc/jit/ir/alias_analysis.h:
```diff
@@ -67,6 +67,8 @@ class AliasDb {
   // hold in memory any element that exists in the other
   TORCH_API bool mayContainAlias(Value* a, Value* b) const;
 
+  TORCH_API bool mayContainAlias(Value* a, const at::ArrayRef<Value*> b) const;
```
Comment?

suo commented on torch/csrc/jit/passes/utils/memory_dag.h:
```diff
@@ -90,6 +90,8 @@ class TORCH_API MemoryDAG {
   // Does `a` hold reference to any memory that is stored in `b`, or vice versa?
   bool mayContainAlias(const Element* a, const Element* b) const;
 
+  bool mayContainAlias(const Element* a, const at::ArrayRef<Element*> b) const;
```
Comment?

suo commented on torch/csrc/jit/ir/alias_analysis.cpp:
```diff
@@ -1333,6 +1333,16 @@ bool AliasDb::mayContainAlias(
+    return false;
+  }
+  auto b_elems = getElements(b);
+  return b_elems.size() == 0
```
is it faster to check this twice vs. delegating the check to MemooryDAG::mayContainAlias?

suo commented:
lgtm

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch][JIT] Support mayContainAlias(Value*, ArrayRef<Value*>)"


We can implement this overload more efficiently.

Differential Revision: [D33039731](https://our.internmc.facebook.com/intern/diff/D33039731/)

[ghstack-poisoned]
